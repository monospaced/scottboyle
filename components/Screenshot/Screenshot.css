@import url('../../theme/vars.css');

.Screenshot {
  visibility: visible;
  max-width: 100%;
  height: auto !important;
  margin: 1.5rem 0;
  transition: filter .2s;
  filter: blur(0);
}

.Screenshot.is-loading {
  visibility: hidden;
  transition: none;
  filter: blur(2px);
}
