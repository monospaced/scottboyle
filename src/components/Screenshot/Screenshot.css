@import url("../../theme/vars.css");

.Screenshot {
  border: 1px solid var(--border-color);
  margin: 1.5rem 0;
  max-width: 100%;
  position: relative;
}

.Screenshot-image {
  bottom: 0;
  filter: blur(2px);
  height: auto !important;
  left: 0;
  max-width: 100%;
  position: absolute;
  right: 0;
  top: 0;
  transition: none;
  visibility: hidden;
}

.Screenshot-shim {
  width: 100%;
}

.Screenshot.is-loaded,
.no-js .Screenshot {
  border-color: transparent;
}

.Screenshot.is-loaded .Screenshot-image,
.no-js .Screenshot-image {
  filter: blur(0);
  transition: filter 0.2s;
  visibility: visible;
}
