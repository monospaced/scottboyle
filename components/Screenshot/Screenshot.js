import React, {Component} from 'react';
import './Screenshot.css';
import {version} from '../../package.json';

class Screenshot extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loaded: true,
    };
    this.imageWillLoad = this.imageWillLoad.bind(this);
  }
  render() {
    const {slug, project: {link, title}} = this.props;
    const image = require(`../../assets/${slug}.png`);
    const img = (
      <img
        className={`Screenshot${!this.state.loaded ? ' is-loading' : ''}`}
        alt={title}
        width="400"
        height={`${(image.width > 398 ? image.height / 2 : image.height) + 2}`}
        src={`/assets/${slug}.png?v=${version}`}
      />
    );
    if (link) {
      return (
        <a href={link}>
          {img}
        </a>
      );
    }
    return img;
  }
  componentDidMount() {
    this.imageWillLoad();
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps !== this.props) {
      this.imageWillLoad(nextProps);
    }
  }
  imageWillLoad(nextProps) {
    const {slug} = nextProps || this.props;
    const img = document.createElement('img');
    this.setState({loaded: false});
    img.onload = () => {
      this.setState({loaded: true});
    };
    img.src = `/assets/${slug}.png?v=${version}`;
  }
}

export default Screenshot;
