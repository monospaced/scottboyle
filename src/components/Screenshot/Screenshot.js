import React, { Component } from "react";

import "./Screenshot.css";

class Screenshot extends Component {
  constructor(props) {
    super(props);
    this.image = React.createRef();
    this.state = { loaded: false };
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    if (nextProps.slug !== prevState.slug) {
      return { loaded: false, slug: nextProps.slug };
    }
    return null;
  }

  render() {
    const { loaded } = this.state;
    const {
      project: {
        image: { height, width },
        link,
        title,
      },
      slug,
    } = this.props;
    const img = (
      <div
        className={`Screenshot${loaded ? " is-loaded" : ""}`}
        style={{ width: `${width}px` }}
      >
        <div
          className="Screenshot-shim"
          style={{ paddingTop: `${(height / width) * 100}%` }}
        />
        <img
          alt={title}
          className="Screenshot-image"
          height={height}
          onLoad={this.imageDidLoad.bind(this)}
          ref={this.image}
          src={require(`../../assets/${slug}.jpg`)}
          width={width}
        />
      </div>
    );

    if (link) {
      return <a href={link}>{img}</a>;
    }
    return img;
  }

  componentDidMount() {
    this.image.current.src = require(`../../assets/${this.props.slug}.jpg`);
  }

  imageDidLoad() {
    this.setState({ loaded: true });
  }
}

export default Screenshot;
