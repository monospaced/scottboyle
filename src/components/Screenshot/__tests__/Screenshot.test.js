import { mount } from "enzyme";
import React from "react";

import data from "../../../scripts/__mocks__/data.js";
import Screenshot from "../Screenshot.js";

describe("Screenshot component", () => {
  const { projects } = data;

  it("should render correctly", () => {
    Object.keys(projects).map(key => {
      const component = mount(
        <Screenshot project={projects[key]} slug={key} />,
      );
      expect(component).toMatchSnapshot();
    });
  });

  it("should handle image load event correctly", () => {
    Object.keys(projects).map(key => {
      const component = mount(
        <Screenshot project={projects[key]} slug={key} />,
      );
      component.instance().imageDidLoad();
      expect(component.state("loaded")).toBe(true);
    });
  });

  it("should handle re-rendering with an identical slug", () => {
    const key = Object.keys(projects)[0];
    const component = mount(<Screenshot project={projects[key]} slug={key} />);
    component.setProps({ slug: key });
    expect(component.state("slug")).toBe(key);
  });
});
