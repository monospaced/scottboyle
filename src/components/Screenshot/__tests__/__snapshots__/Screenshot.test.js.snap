// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot component should render correctly 1`] = `
<Screenshot
  project={
    Object {
      "client": Object {
        "link": "https://monospaced.com",
        "title": "Monospaced",
      },
      "content": "
Building a design system and React component&nbsp;library.
      ",
      "date": "2018",
      "image": Object {
        "height": 275,
        "width": 398,
      },
      "link": "https://www.thomsonreuters.com/en.html",
      "title": "Thomson Reuters",
    }
  }
  slug="thomson-reuters"
>
  <a
    href="https://www.thomsonreuters.com/en.html"
  >
    <div
      className="Screenshot"
      style={
        Object {
          "width": "398px",
        }
      }
    >
      <div
        className="Screenshot-shim"
        style={
          Object {
            "paddingTop": "69.09547738693468%",
          }
        }
      />
      <img
        alt="Thomson Reuters"
        className="Screenshot-image"
        height={275}
        onLoad={[Function]}
        src={
          Object {
            "default": "test-file-stub",
          }
        }
        width={398}
      />
    </div>
  </a>
</Screenshot>
`;

exports[`Screenshot component should render correctly 2`] = `
<Screenshot
  project={
    Object {
      "client": Object {
        "link": "https://monospaced.com",
        "title": "Monospaced",
      },
      "content": "
### Retail Pricing Portal

UI design and complete front-end build, Angular NgRx web&nbsp;app.
      ",
      "date": "2017",
      "image": Object {
        "height": 244,
        "width": 398,
      },
      "link": "",
      "title": "Compass Group",
    }
  }
  slug="compass-group"
>
  <div
    className="Screenshot"
    style={
      Object {
        "width": "398px",
      }
    }
  >
    <div
      className="Screenshot-shim"
      style={
        Object {
          "paddingTop": "61.30653266331658%",
        }
      }
    />
    <img
      alt="Compass Group"
      className="Screenshot-image"
      height={244}
      onLoad={[Function]}
      src={
        Object {
          "default": "test-file-stub",
        }
      }
      width={398}
    />
  </div>
</Screenshot>
`;

exports[`Screenshot component should render correctly 3`] = `
<Screenshot
  project={
    Object {
      "client": Object {
        "link": "https://profitfromplay.com",
        "title": "Profit From Play",
      },
      "content": "
Prototype web e-book reader for digital library startup. Lives on as an open source project:&nbsp;[Paperback](https://monospaced.github.io/paperback/).
      ",
      "date": "2012",
      "image": Object {
        "height": 190,
        "width": 199,
      },
      "link": "",
      "title": "Bilbary",
    }
  }
  slug="bilbary"
>
  <div
    className="Screenshot"
    style={
      Object {
        "width": "199px",
      }
    }
  >
    <div
      className="Screenshot-shim"
      style={
        Object {
          "paddingTop": "95.47738693467338%",
        }
      }
    />
    <img
      alt="Bilbary"
      className="Screenshot-image"
      height={190}
      onLoad={[Function]}
      src={
        Object {
          "default": "test-file-stub",
        }
      }
      width={199}
    />
  </div>
</Screenshot>
`;

exports[`Screenshot component should render correctly 4`] = `
<Screenshot
  project={
    Object {
      "client": Object {
        "link": "https://profitfromplay.com",
        "title": "Profit From Play",
      },
      "content": "
Consulting and advising on a site-wide design refresh. Created indicative templates and assisted with&nbsp;integration.
      ",
      "date": "2012",
      "image": Object {
        "height": 143,
        "width": 199,
      },
      "link": "http://genesreunited.co.uk",
      "title": "Genes Reunited",
    }
  }
  slug="genes-reunited"
>
  <a
    href="http://genesreunited.co.uk"
  >
    <div
      className="Screenshot"
      style={
        Object {
          "width": "199px",
        }
      }
    >
      <div
        className="Screenshot-shim"
        style={
          Object {
            "paddingTop": "71.85929648241206%",
          }
        }
      />
      <img
        alt="Genes Reunited"
        className="Screenshot-image"
        height={143}
        onLoad={[Function]}
        src={
          Object {
            "default": "test-file-stub",
          }
        }
        width={199}
      />
    </div>
  </a>
</Screenshot>
`;

exports[`Screenshot component should render correctly 5`] = `
<Screenshot
  project={
    Object {
      "client": null,
      "content": "
Technical lead on an accessible Web Standards retrofit of key areas of national-lottery.co.uk, including the Lotto and EuroMillions purchase processes, user registration and account&nbsp;management.

Responsible for front-end architecture and all front-end development. Authored company Web Accessibility Guidelines, and conducted usability testing with screen-reader&nbsp;users.
      ",
      "date": "2006",
      "image": Object {
        "height": 284,
        "width": 199,
      },
      "link": "",
      "title": "The National Lottery",
    }
  }
  slug="the-national-lottery"
>
  <div
    className="Screenshot"
    style={
      Object {
        "width": "199px",
      }
    }
  >
    <div
      className="Screenshot-shim"
      style={
        Object {
          "paddingTop": "142.71356783919597%",
        }
      }
    />
    <img
      alt="The National Lottery"
      className="Screenshot-image"
      height={284}
      onLoad={[Function]}
      src={
        Object {
          "default": "test-file-stub",
        }
      }
      width={199}
    />
  </div>
</Screenshot>
`;

exports[`Screenshot component should render correctly 6`] = `
<Screenshot
  project={
    Object {
      "client": Object {
        "link": "",
        "title": "Two Plus Four",
      },
      "content": "
### Age Concern Southwark

Responsible for designing and developing an accessible <abbr title=\\"Disability Discrimination Act\\">DDA</abbr> compliant website for older&nbsp;users.
      ",
      "date": "2005",
      "image": Object {
        "height": 305,
        "width": 199,
      },
      "link": "",
      "title": "Age Concern",
    }
  }
  slug="age-concern"
>
  <div
    className="Screenshot"
    style={
      Object {
        "width": "199px",
      }
    }
  >
    <div
      className="Screenshot-shim"
      style={
        Object {
          "paddingTop": "153.26633165829148%",
        }
      }
    />
    <img
      alt="Age Concern"
      className="Screenshot-image"
      height={305}
      onLoad={[Function]}
      src={
        Object {
          "default": "test-file-stub",
        }
      }
      width={199}
    />
  </div>
</Screenshot>
`;
